src/eddymotion/model/kernels.py

```diff
@@ -0,0 +1,174 @@
+    sigma_sq : float, default=1.0
+        Noise variance term.
+    lambda_s_bounds : pair of floats >= 0 or "fixed", default=(1e-5, 1e4)
+        The lower and upper bound on lambda_s.
+    a_bounds : pair of floats >= 0 or "fixed", default=(1e-5, np.pi)
+        The lower and upper bound on a.
+    sigma_sq_bounds : pair of floats >= 0 or "fixed", default=(1e-5, 1e4).
+    """
+
+    def __init__(
+        self,
+        lambda_s=2.0,
+        a=0.1,
+        sigma_sq=1.0,
+        lambda_s_bounds=(1e-5, 1e4),
+        a_bounds=(1e-5, np.pi),
+        sigma_sq_bounds=(1e-5, 1e4),
+    ):
+        self.lambda_s = lambda_s
+        self.a = a
+        self.sigma_sq = sigma_sq
+        self.lambda_s_bounds = lambda_s_bounds
+        self.a_bounds = a_bounds
+        self.sigma_sq_bounds = sigma_sq_bounds
+
+    @property
+    def hyperparameter_lambda_s(self):
+        return Hyperparameter("lambda_s", "numeric", self.lambda_s_bounds)
+
+    @property
+    def hyperparameter_a(self):
+        return Hyperparameter("a", "numeric", self.a_bounds)
+
+    @property
+    def hyperparameter_sigma_sq(self):
+        return Hyperparameter("sigma_sq", "numeric", self.sigma_sq_bounds)
+
+    def __call__(self, theta, theta_prime=None, eval_gradient=False):
+        """
+        Compute the kernel matrix.
+
+        Parameters
+        ----------
+        theta : array-like of shape (n_samples, n_samples)
+            Precomputed pairwise angles.
+        theta_prime : array-like of shape (n_samples, n_samples), default=None
+            Second input for the kernel function.
+        eval_gradient : bool, default=False
+            Determines whether the gradient with respect to the log of
+            the kernel hyperparameter is computed.
+
+        Returns
+        -------
+        K : array-like of shape (n_samples, n_samples)
+            Kernel matrix.
+        K_gradient : array-like of shape (n_samples, n_samples, n_dims), optional
+            The gradient of the kernel matrix with respect to the log of the
+            hyperparameters. Only returned when `eval_gradient` is True.
+        """
+        if theta_prime is not None:
+            theta = theta_prime
+        K = np.where(theta <= self.a, 1 - 3 * (theta / self.a) ** 2 + 2 * (theta / self.a) ** 3, 0)
+
+        if eval_gradient:
+            K_gradient = np.zeros((theta.shape[0], theta.shape[1], 3))
+            dists_deriv = np.zeros_like(theta)
+            mask = theta <= self.a
+            dists_deriv[mask] = (3 * theta[mask] / self.a**2) - (
+                1.5 * (theta[mask] / self.a) ** 2
+            ) / self.a
+            K_gradient[:, :, 0] = K / self.lambda_s
+            K_gradient[:, :, 1] = self.lambda_s * dists_deriv
+            K_gradient[:, :, 2] = np.eye(len(theta))
+            return self.lambda_s * K + self.sigma_sq * np.eye(len(theta)), K_gradient
+        else:
+            return self.lambda_s * K + self.sigma_sq * np.eye(len(theta))
```

Comment:
The output of the function should be regular:

```suggestion
        K_gradient = None
        if eval_gradient:
            K_gradient = np.zeros((theta.shape[0], theta.shape[1], 3))
            dists_deriv = np.zeros_like(theta)
            mask = theta <= self.a
            dists_deriv[mask] = (3 * theta[mask] / self.a**2) - (
                1.5 * (theta[mask] / self.a) ** 2
            ) / self.a
            K_gradient[:, :, 0] = K / self.lambda_s
            K_gradient[:, :, 1] = self.lambda_s * dists_deriv
            K_gradient[:, :, 2] = np.eye(len(theta))

        return self.lambda_s * K + self.sigma_sq * np.eye(len(theta)), K_gradient
```

Reply:
I reverted this change in [9d0a7c6](https://github.com/nipreps/eddymotion/pull/203/commits/9d0a7c67dfadd26d38b4af8a62e3a1eeb25a609e). Other kernels from the sklearn library don't produce regular output( see[ kernels](https://github.com/scikit-learn/scikit-learn/blob/2621573e6/sklearn/gaussian_process/kernels.py) ), and returning the gradient at any time causes errors when the kernel is called with eval_gradients=False.